src/swar_synthesizer.py

```diff
@@ -4,82 +4,98 @@
 import struct
 import pyaudio
 
-def swar2freq(swar,basefreq):
+SWAR_FREQUENCY_MAPPER = {
+    's':{
+        'shuddh':1,
+        'vakra':1, # Error handling mechanism.
+    },
+    'r':{
+        'shuddh': 1.125,
+        'vakra': 1.058
+    },
+    'g':{
+        'shuddh': 1.266,
+        'vakra': 1.188
+    },
+    'm':{
+        'shuddh': 1.333,
+        'vakra': 1.412
+    },
+    'p':{
+        'shuddh': 1.5,
+        'vakra': 1.5 # Error handling mechanism
+    },
+    'd':{
+        'shuddh': 1.688,
+        'vakra': 1.586
+    },
+    'n':{
+        'shuddh': 1.898,
+        'vakra': 1.78
+    },
+    '0': 0,
+}
+VALID_SWAR_INPUTS_SUBSET_1 = set(['<', '>'])
+VALID_SWAR_INPUTS_SUBSET_2 = set(['s', 'r', 'g', 'm', 'p', 'd', 'n'])
+VALID_SWAR_INPUTS = set(
+    list(VALID_SWAR_INPUTS_SUBSET_1) + list(VALID_SWAR_INPUTS_SUBSET_2)
+)
+
+
+def swar_to_frequency(swar, base_frequency):
     """
-    Contains a dictionary (mapper) that stores relates the swar to 
+    Uses a dictionary (mapper) that relates the swar to
     its defined frequency. I have used the Pythagorean ratios
-    to define the relative frequencies based a fixed basefreq
+    to define the relative frequencies based a fixed base_frequency
     variable that is set to 440 Hz by default, but can be speci-
     fied by the User.
-    This function interprets the octave specifier which halves 
+    This function interprets the octave specifier which halves
     or doubles the value of the defined note to shift down or up
     on the scale.
-    """
-    basefreq_float=float(basefreq) # Perform type conversion once
-
-    mapper={
-        's':{
-            'shuddh':1,
-            'vakra':1, # Error handling mechanism.
-            },
-        'r':{
-            'shuddh': 1.125,
-            'vakra': 1.058
-            },
-        'g':{
-            'shuddh': 1.266,
-            'vakra': 1.188
-            },
-        'm':{
-            'shuddh': 1.333,
-            'vakra': 1.412
-            },
-        'p':{
-            'shuddh': 1.5,
-            'vakra': 1.5 # Error handling mechanism
-            },
-        'd':{
-            'shuddh': 1.688,
-            'vakra': 1.586
-            },
-        'n':{
-            'shuddh': 1.898,
-            'vakra': 1.78
-            },
-        '0':0,
-        
-    }
-
-    """
     For any other input, including captialization, this function
     will specify 0 Hz. Might not be ideal.
     """
-    allowed_inputs=['s','r','g','m','p','d','n','^','<','>']
+    base_frequency_float = float(base_frequency) # Perform type conversion once
+    first_swar = swar[0]
+    is_swar_valid = first_swar in VALID_SWAR_INPUTS
+    if not is_swar_valid:
+        return SWAR_FREQUENCY_MAPPER.get('0')
```

Comment:
Never seen get() before. Learned something new! Thanks!

Reply:
There are two reasons why `dict.get()` is used to fetch data -
* If a key is absent in `dict[key]`, it will raise a `KeyError` which needs to be handled downstream by a `try-except` block.
* You can specify a fallback value -
    For example if the name "Foo" is unavailable in a dict `BABY_NAMES` (not a great baby name, I know 😝), then `BABY_NAMES.get("Foo")` would return `None`. If you have a unique fallback say "Bar", you can specify that like this `BABY_NAMES.get("Foo", "Bar")`. `dict.get(key, fallback_value)` is what is recommended by experienced developers for better readability and debuggability (I think I made this word up).

`dict.get(key, fallback_value)` is the same as
```
value = dict.get(key)
if value is None:
    value = <default_value>
```